shopware/platform
# Add database index for the `order`.`order_number` field

### 1. Why is this change necessary?
Searching for an order by the order_number field in the API or CLI context results in a full table scan.
Sample code:
```
$criteria = (new Criteria())->addFilter(new EqualsFilter('orderNumber', $orderNumber))
    ->addAssociation('transactions');

$this->orderRepository->search($criteria, Context::createDefaultContext())->first();
```

The query takes 2 seconds for 50k orders in the table.

### 2. What does this change do, exactly?

Adds a BTREE index on the ``order.order_number`` field.

### 3. Describe each step to reproduce the issue or behaviour.

1. Create demo orders `bin/console framework:demodata —orders=20000`
2. Run explain for a search query:
  ```
  explain select id
  from `order`
  where order_number='test_order'
  ```
3. Check how many rows have been scanned. -> Full table scan, no indices available.
4. Add an index:
  ```
  ALTER TABLE `order` ADD KEY `idx.order_number` (`order_number`)
  ```
5. Run steps 2 and 3
6. `idx.order_number` is used by the query

### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/workflow/2020-08-03-implement-New-Changelog.md) with all necessary information about my changes
- [x] I have read the contribution requirements and fulfil them.


<a href="https://gitpod.io/#https://github.com/shopware/platform/pull/2875"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
